### PR TITLE
JBPM-6498 - Case file and comments authorisation

### PR DESCRIPTION
**Pull request by mswiderski**

### CODE REVIEW

**Comment by mswiderski** [comment]
Jenkins retest this

**Comment by mswiderski** [comment]
Jenkins retest this

**Comment by mswiderski** [comment]
Jenkins retest this

File: jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/CaseRuntimeDataServiceImpl.java [hunk 1]
```diff
@@ -740,6 +741,27 @@ private void getMilestones(NodeContainer container, Collection<CaseMilestone> re
         return result;
     }
     
+    private Map<String, List<String>> collectDataAccessRestrictions(Process process) {
+        // expected format: item:role1,role2;another:role2,role3
+        // where item and item2 are the case file data item names and role1, role2, role3 are case roles
```
Comment (MarianMacik) [inline jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/CaseRuntimeDataServiceImpl.java]: I cannot see item2 here, just another, so I would update the comment so it is not so confusing.

File: jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/CaseServiceImpl.java [hunk 2]
```diff
@@ -662,5 +724,6 @@ protected boolean isEmpty(Collection<?> data) {
         return false;
     }
 
+
```
Comment (MarianMacik) [inline jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/CaseServiceImpl.java]: If this line is not necessary, can it be removed?

File: jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/CaseRuntimeDataServiceImpl.java (hunk 1 above)
Comment (mswiderski) [reply jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/CaseRuntimeDataServiceImpl.java]: fixed

File: jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/CaseServiceImpl.java (hunk 2 above)
Comment (mswiderski) [reply jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/CaseServiceImpl.java]: removed

File: jbpm-case-mgmt/jbpm-case-mgmt-impl/src/test/java/org/jbpm/casemgmt/impl/CaseServiceImplTest.java [hunk 3]
```diff
@@ -2236,4 +2238,415 @@ public void testStartExpressionCaseWithCaseFile() {
...
+            caseService.addDataToCaseFile(caseId, "dataComplete", true, "owner", "participant");
+
+            activeStages = caseRuntimeDataService.getCaseInstanceStages(caseId, true, new QueryContext());
+            assertNotNull(activeStages);
+            assertEquals(0, activeStages.size());
+            
+            cInstance = caseService.getCaseInstance(caseId, true, false, false, false);
+            assertNotNull(cInstance);
+            assertEquals(FIRST_CASE_ID, cInstance.getCaseId());
+            // mary should not see any data yet
```
Comment (MarianMacik) [inline jbpm-case-mgmt/jbpm-case-mgmt-impl/src/test/java/org/jbpm/casemgmt/impl/CaseServiceImplTest.java]: Just a question, why is there a comment with Mary when we are logged as John?

File: jbpm-case-mgmt/jbpm-case-mgmt-impl/src/test/java/org/jbpm/casemgmt/impl/CaseServiceImplTest.java [hunk 4]
```diff
@@ -2236,4 +2238,415 @@ public void testStartExpressionCaseWithCaseFile() {
...
+            assertEquals(true, caseData.get("dataComplete"));
+            
+            identityProvider.setName("john");
+            
+            caseService.addDataToCaseFile(caseId, "anotherDataItem", "first version");
+            
+            cInstance = caseService.getCaseInstance(caseId, true, false, false, false);
+            assertNotNull(cInstance);
+            assertEquals(FIRST_CASE_ID, cInstance.getCaseId());
+            // mary should not see any data yet
```
Comment (MarianMacik) [inline jbpm-case-mgmt/jbpm-case-mgmt-impl/src/test/java/org/jbpm/casemgmt/impl/CaseServiceImplTest.java]: Same as before - Mary vs John comment.

File: jbpm-case-mgmt/jbpm-case-mgmt-impl/src/test/java/org/jbpm/casemgmt/impl/CaseServiceImplTest.java [hunk 5]
```diff
@@ -2236,4 +2238,415 @@ public void testStartExpressionCaseWithCaseFile() {
...
+            assertEquals("main street 10, NYC", caseData.get("contactInfo"));
+            assertEquals(true, caseData.get("dataComplete"));
+            assertEquals("first version", caseData.get("anotherDataItem"));
+            
+            caseService.removeDataFromCaseFile(caseId, "contactInfo");
+            
+            cInstance = caseService.getCaseInstance(caseId, true, false, false, false);
+            assertNotNull(cInstance);
+            assertEquals(FIRST_CASE_ID, cInstance.getCaseId());
+            // mary should not see any data yet
```
Comment (MarianMacik) [inline jbpm-case-mgmt/jbpm-case-mgmt-impl/src/test/java/org/jbpm/casemgmt/impl/CaseServiceImplTest.java]: Same as before - Mary vs John comment.

File: jbpm-case-mgmt/jbpm-case-mgmt-impl/src/test/java/org/jbpm/casemgmt/impl/CaseServiceImplTest.java [hunk 6]
```diff
@@ -2236,4 +2238,415 @@ public void testStartExpressionCaseWithCaseFile() {
...
+                
+                // mary is not allowed to removed comments that she has no role to
+                assertTrue(e.getMessage().contains("User mary does not have access to data"));
+            }
+            
+            identityProvider.setName("john");
+            cInstance = caseService.getCaseInstance(caseId, true, false, false, false);
+            assertNotNull(cInstance);
+            assertEquals(FIRST_CASE_ID, cInstance.getCaseId());
+            // mary should not see any data yet
```
Comment (MarianMacik) [inline jbpm-case-mgmt/jbpm-case-mgmt-impl/src/test/java/org/jbpm/casemgmt/impl/CaseServiceImplTest.java]: Same as before - Mary vs John comment.

File: jbpm-case-mgmt/jbpm-case-mgmt-impl/src/test/java/org/jbpm/casemgmt/impl/CaseServiceImplTest.java (hunk 3 above)
Comment (mswiderski) [reply jbpm-case-mgmt/jbpm-case-mgmt-impl/src/test/java/org/jbpm/casemgmt/impl/CaseServiceImplTest.java]: because I really like Mary :)

Comment (MarianMacik) [reply jbpm-case-mgmt/jbpm-case-mgmt-impl/src/test/java/org/jbpm/casemgmt/impl/CaseServiceImplTest.java]: I mean I understand that as a John you check that he sees all items and then you log in as Mary and there is a comment "mary should not see any data yet" - so she sees only one item, contact info is hidden for her. But why is this comment there when we check if john sees everything - I am clearly missing something :)

File: jbpm-case-mgmt/jbpm-case-mgmt-impl/src/test/java/org/jbpm/casemgmt/impl/CaseServiceImplTest.java [hunk 7]
```diff
@@ -2236,4 +2238,415 @@ public void testStartExpressionCaseWithCaseFile() {
...
+            
+            // mary is not the owner so should not see the comment that is only for role owner role
+            identityProvider.setName("mary");
+            caseComments = caseService.getCaseComments(FIRST_CASE_ID, new QueryContext());
+            assertNotNull(caseComments);
+            assertEquals(0, caseComments.size());
+            
+            identityProvider.setName("john");
+
+            caseService.updateCaseComment(FIRST_CASE_ID, comment.getId(), comment.getAuthor(), "Updated " + comment.getComment(), "participant", "owner");
```
[the remark this answers is not in the document]
Comment (mswiderski) [reply jbpm-case-mgmt/jbpm-case-mgmt-impl/src/test/java/org/jbpm/casemgmt/impl/CaseServiceImplTest.java]: added negative path for the update comments

File: jbpm-case-mgmt/jbpm-case-mgmt-impl/src/test/java/org/jbpm/casemgmt/impl/CaseServiceImplTest.java (hunk 3 above)
Comment (mswiderski) [reply jbpm-case-mgmt/jbpm-case-mgmt-impl/src/test/java/org/jbpm/casemgmt/impl/CaseServiceImplTest.java]: it was simply copy paste thingy

**Pull request merged**